log: Log File size limit and Log rotation for fluent-bit logs

phiremande opened the pull request:
Currently fluent-bit logs can be redirected to a file using Log_File.
But there is no configuration option to limit the size of the file. Additionally, there is no log rotation that can be enabled for fluent-bit logs.
These might be useful when we want to collect fluent-bit process logs itself and also to debug.

Added new configuration parameters:
Log_File_Size - Size limit in MB for the file specified by Log_File.
Log_File_History - Max number of backups to be kept.
These backups are kept as <Log_File>.1, <Log_File>.2 etc.

Sample configuration.
[SERVICE]
    Flush        1
    Daemon       Off
    Log_Level    debug
    Log_File     ./fb_log.log
    Log_File_Size     1   
    Log_File_History  2

This PR has the corresponding changes.

phiremande committed:
log: changes for limiting/rotating fluent-bit logs

Signed-off-by: Pradeep Hiremande <16595434+phiremande@users.noreply.github.com>

edsiper commented on src/flb_log.c:
```diff
@@ -57,10 +58,78 @@ static inline int consume_byte(flb_pipefd_t fd)
+
+    sz_mb = size / (1024.00 * 1024.00);
+    if (sz_mb >= log->max_sz_mb) {
+        return true;
```
use FLB_TRUE/FLB_FALSE instead (and change return type to int)

phiremande replied:
updated.

edsiper commented on src/flb_log.c:
```diff
@@ -57,10 +58,78 @@ static inline int consume_byte(flb_pipefd_t fd)
+    int ret = -1;
+
+    if (log->max_history == 0) {
+        //no backups.
```
use /* ... */ for comments

phiremande replied:
updated

edsiper commented on src/flb_log.c:
```diff
@@ -57,10 +58,78 @@ static inline int consume_byte(flb_pipefd_t fd)
+    //max 2 digits for backup name suffix and 1 each for '.' and '\0'.
+    backup_file_name = flb_malloc(strlen(log->out) + 4);
+    if (!backup_file_name) {
+        perror("malloc");
```
use flb_errno()

phiremande replied:
changed

edsiper commented on src/flb_log.c:
```diff
@@ -57,10 +58,78 @@ static inline int consume_byte(flb_pipefd_t fd)
+    }
+
+    //max 2 digits for backup name suffix and 1 each for '.' and '\0'.
+    backup_file_name = flb_malloc(strlen(log->out) + 4);
```
set the malloc size in a variable before malloc

phiremande replied:
updated.

edsiper commented on src/flb_log.c:
```diff
@@ -57,10 +58,78 @@ static inline int consume_byte(flb_pipefd_t fd)
+        perror("malloc");
+        return;
+    }
+    src_file_name = flb_malloc(strlen(log->out) + 4);
```
dont nest strlen, set the malloc size in a variable

phiremande replied:
updated

edsiper commented on src/flb_log.c:
```diff
@@ -57,10 +58,78 @@ static inline int consume_byte(flb_pipefd_t fd)
+    }
+    src_file_name = flb_malloc(strlen(log->out) + 4);
+    if (!src_file_name) {
+        perror("malloc");
```
flb_errno();

phiremande replied:
updated

edsiper commented on src/flb_log.c:
```diff
@@ -57,10 +58,78 @@ static inline int consume_byte(flb_pipefd_t fd)
+        return;
+    }
+
+    for (int l = log->max_history; l > 1; l--) {
```
do not declare variables inside the ```for``` statement

edsiper commented on src/flb_log.c:
```diff
@@ -57,10 +58,78 @@ static inline int consume_byte(flb_pipefd_t fd)
+    }
+
+    for (int l = log->max_history; l > 1; l--) {
+        sprintf(backup_file_name, "%s.%d", log->out, l);
```
use snprintf instead

phiremande replied:
updated

edsiper commented on src/flb_log.c:
```diff
@@ -57,10 +58,78 @@ static inline int consume_byte(flb_pipefd_t fd)
+        flb_free(src_file_name);
+    }
+
+    sprintf(backup_file_name, "%s.1", log->out);
```
use snprintf

phiremande replied:
updated

phiremande commented:
@edsiper , kindly let me know if you have any further suggestions.

phiremande commented:
@edsiper , please let me know if this PR still makes sense and are there any plans to merge this.

edsiper commented on src/flb_log.c:
```diff
@@ -57,10 +57,82 @@ static inline int consume_byte(flb_pipefd_t fd)
+        return;
+    }
+
+    int l;
```
let's move this to the top level of the function and use `i` instead

phiremande replied:
done

edsiper commented on src/flb_log.c:
```diff
@@ -57,10 +57,82 @@ static inline int consume_byte(flb_pipefd_t fd)
+        snprintf(src_file_name, filename_len, "%s.%d", log->out, l-1);
+        /* since rename behaviour is implementation dependent when new file exists, remove the last file. */
+        if (l == log->max_history) {
+            remove(backup_file_name);
```
we might need a Windows compat version for this. cc: @leonardo-albertovich

phiremande replied:
remove is from stdio.h and hence shouldn't be an issue I believe.

leonardo-albertovich replied:
It is according to MSDN but I couldn't find / verify when was it introduced (Sorry windows 95 users!)

edsiper commented on src/flb_log.c:
```diff
@@ -73,7 +145,16 @@ static inline int log_push(struct log_message *msg, struct flb_log *log)
         }
         ret = write(fd, msg->msg, msg->size);
+        /* Let's check the size before closing as we have open file descriptor. */
+        fstat(fd, &buf);
```
check the return status of fstat()

phiremande replied:
done

edsiper reviewed (changes requested):
thanks for this PR.

I added some minor change requests before to get this merged we need to sort out two things:

- rebase on top of GIT master
- all I/O functions must work on Linux and Windows, .eg: remove(3) is not valid on windows

phiremande commented:
> thanks for this PR.
> 
> I added some minor change requests before to get this merged we need to sort out two things:
> 
>     * rebase on top of GIT master
> 
>     * all I/O functions must work on Linux and Windows, .eg: remove(3) is not valid on windows

Kindly let me know if the changes look fine now. If you have any more suggestions please do let me know. Also should I use "log" or "engine" in the commit message?

patrick-stephens commented:
You may need a rebase to solve the fuzzing failures but the unit tests also need resolving. There seems to be quite a few failures which will prevent merging.

phiremande commented:
> You may need a rebase to solve the fuzzing failures but the unit tests also need resolving. There seems to be quite a few failures which will prevent merging.

@patrick-stephens , I have rebased and fixed UT failures. Can you please let me know what should be the commit message? I guess it is failing due to the component missing in commit message.

patrick-stephens commented:
Yeah you need it in the format: `component: message `

phiremande commented:
@patrick-stephens , please do let me know if there is anything else pending from my side.

phiremande commented:
@edsiper , I have addressed your comments. Please do let me know if you still have any concerns.

patrick-stephens commented:
@phiremande did you set up the docs PR as well? Can you link it so we can merge that once this is merged?

leonardo-albertovich commented on src/flb_log.c:
```diff
@@ -62,10 +62,83 @@ static inline int consume_byte(flb_pipefd_t fd)
+        ret = rename(src_file_name, backup_file_name);
+        if (ret != 0) {
+            /* we could not rename; delete the source file. */
+            remove(src_file_name);
```
I think removing on error might not be what our users expect and that we might instead want to implement a fallback mechanism such as renaming the next file with a temporary filename (eg. same filename with a trailing dash) and try to normalize the situation on the next iteration and if it doesn't then set a flag to stop rotation altogether while also notifying the user of the situation.

I don't think silently discarding logs is something most users would like.

leonardo-albertovich commented on src/flb_log.c:
```diff
@@ -62,10 +62,83 @@ static inline int consume_byte(flb_pipefd_t fd)
+            remove(src_file_name);
+        }
+    }
+    if (src_file_name) {
```
Can we move this block to line 130 and remove the if blocks? if we got there it means the pointers are or at the very least should be, we could check them and report if they are not but I think it might not really be all that useful.

leonardo-albertovich commented on src/flb_log.c:
```diff
@@ -62,10 +62,83 @@ static inline int consume_byte(flb_pipefd_t fd)
+        return;
+    }
+
+    for (i = log->max_history; i > 1; i--) {
```
I think we might want to refactor this function, if we set the history limit to 200 it will try to rename 200 files each time which with this code structure is necessary to keep the code from messing up preexisting rotated files.

leonardo-albertovich commented on src/flb_log.c:
```diff
@@ -62,10 +62,83 @@ static inline int consume_byte(flb_pipefd_t fd)
+    ret = rename(log->out, backup_file_name);
+    if (ret != 0) {
+        /* we could not rename; delete the source file. */
+        remove(log->out);
```
Same as 116, also, we need to call `flb_errno` in these cases.

leonardo-albertovich commented on src/flb_log.c:
```diff
@@ -62,10 +62,83 @@ static inline int consume_byte(flb_pipefd_t fd)
+    int status = -1;
+    off_t size;
+    struct stat buf;
+    int backup = FLB_FALSE;
```
Let's call this `rotation_required` or something like meaningful like that.

leonardo-albertovich commented on src/flb_log.c:
```diff
@@ -78,7 +151,21 @@ static inline int log_push(struct log_message *msg, struct flb_log *log)
         }
         ret = write(fd, msg->msg, msg->size);
+        /* Let's check the size before closing as we have open file descriptor. */
+        status = fstat(fd, &buf);
```
Do we want to issue an `fstat` call each time a log line is pushed? I guess this is related to opening and closing the file each time as well which is something I am slightly against. If we decide that we'll keep on opening and closing the file each time I guess that overhead is acceptable, however, I'd really like us to change it if possible.
Of course keeping an internal file size counter could lead to slight deviations if the file was externally modified (which shouldn't really happen) but even then we could make a verification and re-sync call before rotating the file to catch external rotations if the user messes up the configuration (which we should communicate).

What do you think @edsiper?

edsiper commented:
@patrick-stephens can you pls verify functionality of this new feature ?

patrick-stephens commented:
Will do, should this PR cover Windows & Linux functionality so it works on both?

edsiper commented:
@patrick-stephens both ideally

qihonggang commented:
@patrick-stephens When will this PR merge into the master branch？

patrick-stephens commented:
Apologies I've had little time to look at this so far but will endeavour to soon.
It can't be merged right now anyway as it has a conflict.

said1995 commented:
Hi Team,
Any update on this request?
This would be very helpful.

patrick-stephens commented:
Apologies, I've still not tested this yet.

ustndagsemih commented:
Hi team,
Any thoughts on when would this be merged?

robert-heinzmann-logmein commented:
Any update on this one, I think it would be very valuable to have predictable log sizes for fluent bit itself.

wojciechtobis commented:
Hey @phiremande, are you going to finish this PR? I can continue to work on it as I really need these changes.

hcdmonitor commented:
Hey everyone, I'm grateful for the work that has already been carried out on this issue. 
The log file size limit and rotation would be very useful to us to prevent disk utilisation issues caused by fluent-bit.

Is there an update as to when the changes will be merged?

patrick-stephens commented:
> Hey everyone, I'm grateful for the work that has already been carried out on this issue. The log file size limit and rotation would be very useful to us to prevent disk utilisation issues caused by fluent-bit.
> 
> Is there an update as to when the changes will be merged?

It needs rebasing and updating so if you pick it up @hcdmonitor it may speed things up.

etolbakov commented on src/flb_log.c:
```diff
@@ -62,10 +62,83 @@ static inline int consume_byte(flb_pipefd_t fd)
+    }
+
+    /* max 2 digits for backup name suffix and 1 each for '.' and '\0' */
+    size_t filename_len = strlen(log->out) + 4;
```
is this expected here?
```suggestion
    size_t filename_len = snprintf(NULL, 0, "%"PRIu64, log->out);
```

etolbakov commented on src/flb_log.c:
```diff
@@ -62,10 +62,83 @@ static inline int consume_byte(flb_pipefd_t fd)
+    }
+    src_file_name = flb_malloc(filename_len);
+    if (!src_file_name) {
+        flb_errno();
```
is this expected here?
```suggestion
        flb_free(backup_file_name);
        flb_errno();
```

etolbakov commented:
Hello everyone, 
I'm also interested in this feature! 
Am I correct assuming that the only way of rebasing here is to open a new PR @patrick-stephens @hcdmonitor ?
I don't think anyone except @phiremande has permissions to push to `phiremande:log-rotate` branch. 
Please let me know if I'm wrong.

ashraf133 commented:
Hello, any news about this feature ? it is fundamental

alanbrito commented:
This is a very basic feature. We cannot go live with a component that will almost guarantee and outage eventually, due to filling the disk.

Is there any progress?

patrick-stephens commented:
> Hello everyone, I'm also interested in this feature! Am I correct assuming that the only way of rebasing here is to open a new PR @patrick-stephens @hcdmonitor ? I don't think anyone except @phiremande has permissions to push to `phiremande:log-rotate` branch. Please let me know if I'm wrong.

It may be easier with a new PR, you can generate the patch for this one by just adding a `.diff` or `.patch` to this one.

For everyone asking about log rotate, typically I would do this separately via `logrotate` or something central to the OS rather than relying on each individual process to do it with all the associated configs to then manage too.

pkqsun commented:
Hello, any news about this feature? 
I really need this feature on my Windows machines.
